File: go/0098M-验证二叉搜索树.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	ans := make([]int, 0)
	var inOrder func(root *TreeNode)
	inOrder = func(root *TreeNode) {
		if root == nil {
			return
		}
		inOrder(root.Left)
		ans = append(ans, root.Val)
		inOrder(root.Right)
	}
	inOrder(root)
	for i := 1; i < len(ans); i++ {
		if ans[i] <= ans[i-1] {
			return false
		}
	}
	return true
}

File: cpp/0098M-验证二叉搜索树.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-22 10:19:26
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    TreeNode* pre = nullptr;  // 记录左边的结点

    bool isValidBST(TreeNode* root) {  // 中序遍历递增
        if (root == nullptr) return true;
        bool left = isValidBST(root->left);
        if (pre != nullptr && pre->val >= root->val) return false;
        pre = root;
        bool right = isValidBST(root->right);
        return left && right;
    }
};
